theano/gof/opt.py

```diff
@@ -2946,6 +2947,24 @@ def copy_stack_trace(from_var, to_var):
         # Copy over stack traces from from_var to each variable to
         # to_var, including the stack_trace of the to_var before
         to_var.tag.trace = getattr(to_var.tag, 'trace', []) + tr
+    return to_var
+
+
+@contextlib.contextmanager
+def inherit_stack_trace(from_var):
+    """
+    Contextmanager that copies the stack trace from one or more tensor variables to all tensor
+    variables constructed in the body.
+
+    Parameters
+    ----------
+    from_var
+        Tensor variable or list of tensor variables to copy stack traces from.
+
+    """
+    with graph.nodes_constructed() as new_nodes:
```

Comment:
Having a context manager inside a context manager make it more complicated to understand how this all work. We can keep it like this, but I think it should be explained somewhere in a follow up PR.

Reply:
Yes, exactly. I would try to write a more readable form of this.